Crash when gathering diagnostic information on WAS8.5

Hello,

I got this with an application on the current version of SEED on a WAS8.5 server after that the application throw an uncaught exception :

    [7/22/15 17:39:08:211 CEST] 0000007c SystemOut     O   [2015-07-22 17:39:08,210] ERROR [o.s.seed.core.internal.DiagnosticManagerImpl : 85] - Unable to write diagnostic information
    org.seedstack.seed.core.api.SeedException: (COREUTILS) Method invocation failed
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.7.0]
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:86) ~[na:1.7.0]
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:58) ~[na:1.7.0]
        at java.lang.reflect.Constructor.newInstance(Constructor.java:542) ~[na:2.6 (12-15-2014)]
        at org.seedstack.seed.core.api.SeedException.wrap(SeedException.java:435) ~[seed-core-support-specs-2.0.0-M3.jar:na]
        at org.seedstack.seed.core.api.SeedException.wrap(SeedException.java:419) ~[seed-core-support-specs-2.0.0-M3.jar:na]
        at org.seedstack.seed.core.utils.SeedReflectionUtils.invokeMethod(SeedReflectionUtils.java:587) ~[seed-core-support-core-2.0.0-M3.jar:na]
        at org.seedstack.seed.web.internal.WebDiagnosticCollector.buildFilterList(WebDiagnosticCollector.java:70) ~[seed-web-support-core-2.0.0-M3.jar:na]
        at org.seedstack.seed.web.internal.WebDiagnosticCollector.collect(WebDiagnosticCollector.java:36) ~[seed-web-support-core-2.0.0-M3.jar:na]
        at org.seedstack.seed.core.internal.DiagnosticManagerImpl.collectAllDiagnostics(DiagnosticManagerImpl.java:110) ~[seed-core-support-core-2.0.0-M3.jar:na]
        at org.seedstack.seed.core.internal.DiagnosticManagerImpl.dumpDiagnosticReport(DiagnosticManagerImpl.java:83) ~[seed-core-support-core-2.0.0-M3.jar:na]
        at org.seedstack.seed.core.internal.CorePlugin$3.uncaughtException(CorePlugin.java:79) [seed-core-support-core-2.0.0-M3.jar:na]
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:845) [na:1.7.0]
        at java.lang.ThreadGroup.uncaughtException(ThreadGroup.java:843) [na:1.7.0]
        at java.lang.Thread.uncaughtException(Thread.java:1384) [na:1.7.0]
    Caused by: java.lang.NoSuchMethodException: com.ibm.ws.webcontainer.filter.FilterConfig.getInitParameters()
        at java.lang.Class.throwNoSuchMethodException(Class.java:446) ~[na:1.7.0]
        at java.lang.Class.getDeclaredMethod(Class.java:879) ~[na:1.7.0]
        at org.seedstack.seed.core.utils.SeedReflectionUtils.invokeMethod(SeedReflectionUtils.java:583) ~[seed-core-support-core-2.0.0-M3.jar:na]
        ... 8 common frames omitted

I think that getInitParameters() is probably in an inherited class of com.ibm.ws.webcontainer.filter.FilterConfig and since SeedReflectionsUtils.invokeMethod(...) doesn't look in super class it fails to find the method.

--- a/web-support/core/src/main/java/org/seedstack/seed/web/internal/WebDiagnosticCollector.java
+++ b/web-support/core/src/main/java/org/seedstack/seed/web/internal/WebDiagnosticCollector.java
@@ -9,14 +9,18 @@
  */
 package org.seedstack.seed.web.internal;
 
-import com.google.common.collect.Lists;
-import org.seedstack.seed.core.spi.diagnostic.DiagnosticInfoCollector;
-
-import javax.servlet.ServletContext;
+import java.util.ArrayList;
+import java.util.Collection;
 import java.util.HashMap;
 import java.util.Map;
 
-import static org.seedstack.seed.core.utils.SeedReflectionUtils.invokeMethod;
+import javax.servlet.FilterRegistration;
+import javax.servlet.ServletContext;
+import javax.servlet.ServletRegistration;
+
+import org.seedstack.seed.core.spi.diagnostic.DiagnosticInfoCollector;
+
+import com.google.common.collect.Lists;
 
 class WebDiagnosticCollector implements DiagnosticInfoCollector {
     private final ServletContext servletContext;
@@ -30,47 +34,55 @@ class WebDiagnosticCollector implements DiagnosticInfoCollector {
         Map<String, Object> result = new HashMap<String, Object>();
 
         result.put("version", String.format("%d.%d", servletContext.getMajorVersion(), servletContext.getMinorVersion()));
-        result.put("effective-version", String.format("%d.%d", invokeMethod(servletContext, "getEffectiveMajorVersion"), invokeMethod(servletContext, "getEffectiveMinorVersion")));
-        result.put("servlets", buildServletList());
-        result.put("filters", buildFilterList());
-
+        if (servletContext.getMajorVersion() >= 3) {
+            result.put("effective-version",
+                    String.format("%d.%d", servletContext.getEffectiveMajorVersion(), servletContext.getEffectiveMinorVersion()));
+            result.put("servlets", buildServletList());
+            result.put("filters", buildFilterList());
+        }
         return result;
     }
 
-    @SuppressWarnings("unchecked")
     private Map<String, Object> buildServletList() {
         Map<String, Object> servletMap = new HashMap<String, Object>();
-        if (servletContext.getMajorVersion() >= 3) {
-            for (Map.Entry<String, ?> servletRegistrationEntry : ((Map<String, ?>) invokeMethod(servletContext, "getServletRegistrations")).entrySet()) {
-                Object servletRegistration = servletRegistrationEntry.getValue();
-                Map<String, Object> servletRegistrationInfo = new HashMap<String, Object>();
+        for (Map.Entry<String, ? extends ServletRegistration> servletRegistrationEntry : servletContext.getServletRegistrations().entrySet()) {
+            ServletRegistration servletRegistration = servletRegistrationEntry.getValue();
+            Map<String, Object> servletRegistrationInfo = new HashMap<String, Object>();
 
-                servletRegistrationInfo.put("class", invokeMethod(servletRegistration, "getClassName"));
-                servletRegistrationInfo.put("parameters", invokeMethod(servletRegistration, "getInitParameters"));
-                servletRegistrationInfo.put("mappings", Lists.newArrayList(invokeMethod(servletRegistration, "getMappings")));
-
-                servletMap.put(servletRegistrationEntry.getKey(), servletRegistrationInfo);
+            servletRegistrationInfo.put("class", servletRegistration.getClassName());
+            servletRegistrationInfo.put("parameters", servletRegistration.getInitParameters());
+            Collection<String> mappings = servletRegistration.getMappings();
+            if (mappings == null) {
+                mappings = new ArrayList<String>();
             }
+            servletRegistrationInfo.put("mappings", Lists.newArrayList(mappings));
+
+            servletMap.put(servletRegistrationEntry.getKey(), servletRegistrationInfo);
         }
 
         return servletMap;
     }
 
-    @SuppressWarnings("unchecked")
     private Map<String, Object> buildFilterList() {
         Map<String, Object> filterMap = new HashMap<String, Object>();
-        if (servletContext.getMajorVersion() >= 3) {
-            for (Map.Entry<String, ?> filterRegistrationEntry : ((Map<String, ?>) invokeMethod(servletContext, "getFilterRegistrations")).entrySet()) {
-                Object filterRegistration = filterRegistrationEntry.getValue();
-                Map<String, Object> filterRegistrationInfo = new HashMap<String, Object>();
+        for (Map.Entry<String, ? extends FilterRegistration> filterRegistrationEntry : servletContext.getFilterRegistrations().entrySet()) {
+            FilterRegistration filterRegistration = filterRegistrationEntry.getValue();
+            Map<String, Object> filterRegistrationInfo = new HashMap<String, Object>();
 
-                filterRegistrationInfo.put("class", invokeMethod(filterRegistration, "getClassName"));
-                filterRegistrationInfo.put("parameters", invokeMethod(filterRegistration, "getInitParameters"));
-                filterRegistrationInfo.put("servlet-name-mappings", Lists.newArrayList(invokeMethod(filterRegistration, "getServletNameMappings")));
-                filterRegistrationInfo.put("url-pattern-mappings", Lists.newArrayList(invokeMethod(filterRegistration, "getUrlPatternMappings")));
-
-                filterMap.put(filterRegistrationEntry.getKey(), filterRegistrationInfo);
+            filterRegistrationInfo.put("class", filterRegistration.getClassName());
+            filterRegistrationInfo.put("parameters", filterRegistration.getInitParameters());
+            Collection<String> mappings = filterRegistration.getServletNameMappings();
+            if (mappings == null) {
+                mappings = new ArrayList<String>();
             }
+            filterRegistrationInfo.put("servlet-name-mappings", Lists.newArrayList(mappings));
+            Collection<String> patterns = filterRegistration.getUrlPatternMappings();
+            if (patterns == null) {
+                patterns = new ArrayList<String>();
+            }
+            filterRegistrationInfo.put("url-pattern-mappings", Lists.newArrayList(patterns));
+
+            filterMap.put(filterRegistrationEntry.getKey(), filterRegistrationInfo);
         }
 
         return filterMap;

